rootston/wl_shell.c

```diff
@@ -10,17 +10,30 @@
 #include "rootston/server.h"
 #include "rootston/input.h"
 
+static void handle_destroy(struct wl_listener *listener, void *data) {
+	struct roots_wl_shell_surface *roots_surface =
+		wl_container_of(listener, roots_surface, destroy);
+	wl_list_remove(&roots_surface->destroy.link);
+	wl_list_remove(&roots_surface->ping_timeout.link);
+	view_destroy(roots_surface->view);
+	free(roots_surface);
+}
+
 void handle_wl_shell_surface(struct wl_listener *listener, void *data) {
 	struct roots_desktop *desktop =
 		wl_container_of(listener, desktop, wl_shell_surface);
 
 	struct wlr_wl_shell_surface *surface = data;
-	wlr_log(L_DEBUG, "new wl_shell surface");
-	//wlr_wl_shell_surface_ping(surface);
+	wlr_log(L_DEBUG, "new shell surface: title=%s, class=%s",
+		surface->title, surface->class_);
+	//wlr_wl_shell_surface_ping(surface); // TODO: segfaults
```

Comment:
FWIW I think you could implement the whole ping/pong workflow entirely in wlr_wl_shell, no need to expose this function (though I understand this wasn't part of your changes).

Reply:
Hmm, I had the feeling compositors might want to call `wlr_wl_shell_surface_ping` to check that a surface is responsive. When is the compositor supposed to call ping? Also, this is exposed in xdg shell too. Should we remove it from xdg shell and wl_shell?

Reply:
We can instead expose an event for when we believe the surface has become unresponsive.